### PR TITLE
IR-1652 / IR-1986 Interactable Input Refactor

### PR DESCRIPTION
**Pull request by the-reality-engineer**
## Summary

IR-1652
IR-1986

ClientInputSystem - Routes input from InputSourceComponent, through InputComponent, to InputSinkComponent
  1. Each InputSource raycasts into the scene
  2. The hit results (meshes and/or rigidbodies) are traversed upwards to find the closest InputComponent
  3. The InputComponent explicitly routes input to itself and/or another entity

Various new hooks that allow entities to easily consume routed input:

`InputComponent.useExecuteWithInput(executeOnInput:(inputEid:Entity) => {}) : Entity`
 - walks up the entity tree, finding closest InputSinkComponent, and grabs inputEid from that

` InputComponent.useHasFocus()`
 - returns true if the entity input context has inputSources array length of non-zero.

### CODE REVIEW

**Review by SamMazerIR — APPROVED** [review]
functional input refactor, allows sending input to arbitrary entities. InteractableComponent can have its UI management broken out later if need be. Grabbables need network authority fixes before they're working again

File: packages/spatial/src/input/properties/InputComponentNodeEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import React, { useEffect } from 'react'
+import { useTranslation } from 'react-i18next'
+import { InputComponent } from '../components/InputComponent'
+
+type OptionsType = Array<{
+  label: string
+  value: string
+}>
+
+export const InputComponentNodeEditor: EditorComponentType = (props) => {
```
Comment (HexaField) [inline packages/spatial/src/input/properties/InputComponentNodeEditor.tsx:52]: this should not be in the spatial package, it should be in the editor package. might crash builder or servers.

File: packages/spatial/src/input/systems/ClientInputSystem.tsx [hunk 2]
```diff
@@ -38,20 +38,24 @@ import {
...
 } from '@etherealengine/ecs/src/ComponentFunctions'
 import { Engine } from '@etherealengine/ecs/src/Engine'
-import { Entity } from '@etherealengine/ecs/src/Entity'
+import { Entity, UndefinedEntity } from '@etherealengine/ecs/src/Entity'
 import { createEntity, removeEntity, useEntityContext } from '@etherealengine/ecs/src/EntityFunctions'
 import { defineQuery, QueryReactor } from '@etherealengine/ecs/src/QueryFunctions'
 import { defineSystem } from '@etherealengine/ecs/src/SystemFunctions'
 import { InputSystemGroup, PresentationSystemGroup } from '@etherealengine/ecs/src/SystemGroups'
-import { getState, useMutableState } from '@etherealengine/hyperflux'
+import { AvatarComponent } from '@etherealengine/engine/src/avatar/components/AvatarComponent'
```
Comment (HexaField) [inline packages/spatial/src/input/systems/ClientInputSystem.tsx:46]: this breaks our import hierarchy

File: packages/spatial/src/input/systems/ClientInputSystem.tsx [hunk 3]
```diff
@@ -119,11 +123,22 @@ const pointers = defineQuery([InputPointerComponent, InputSourceComponent, Not(X
...
+
 const inputXRUIs = defineQuery([InputComponent, VisibleComponent, XRUIComponent])
 const inputBoundingBoxes = defineQuery([InputComponent, VisibleComponent, BoundingBoxComponent])
 const inputObjects = defineQuery([InputComponent, VisibleComponent, GroupComponent])
+const spatialInputObjects = defineQuery([
+  InputComponent,
+  VisibleComponent,
+  TransformComponent,
+  Not(CameraComponent),
+  Not(AvatarComponent)
```
Comment (HexaField) [inline packages/spatial/src/input/systems/ClientInputSystem.tsx:140]: we need to find a different way of handling this, avatars must not be included in the spatial package

File: packages/spatial/src/input/systems/ClientInputSystem.tsx [hunk 4]
```diff
@@ -188,87 +205,138 @@ const execute = () => {
...
+    if (
+      capturedEntity.value === UndefinedEntity &&
+      sortedIntersections.length === 0 &&
+      !hasComponent(sourceEid, InputPointerComponent)
+    ) {
+      let closestEntity = UndefinedEntity
+      let closestDistanceSquared = Infinity
+
+      //use sourceEid if controller (one InputSource per controller), otherwise use avatar rather than InputSource-emulated-pointer
+      const selfAvatarEntity = AvatarComponent.getSelfAvatarEntity()
```
Comment (HexaField) [inline packages/spatial/src/input/systems/ClientInputSystem.tsx:294]: again, AvatarComponent must not be referenced in the spatial package

File: packages/spatial/src/input/systems/ClientInputSystem.tsx [hunk 5]
```diff
@@ -399,7 +474,8 @@ const usePointerInputSources = () => {
     const pointerEnter = (event: PointerEvent) => {
       setComponent(emulatedInputSourceEntity, InputPointerComponent, {
         pointerId: event.pointerId,
-        canvasEntity: canvasEntity
+        canvasEntity: canvasEntity,
+        cameraEntity: canvasEntity //TODO likely want to double check this is done with a guaranteed camera (though it should exist)
```
Comment (HexaField) [inline packages/spatial/src/input/systems/ClientInputSystem.tsx:478]: in what situation would we have a camera that is not the same as the canvas?

File: packages/spatial/src/input/properties/InputComponentNodeEditor.tsx (hunk 1 above)
Comment (SamMazerIR) [reply packages/spatial/src/input/properties/InputComponentNodeEditor.tsx:52]: moved, thanks for the reminder!

File: packages/spatial/src/input/systems/ClientInputSystem.tsx (hunk 5 above)
Comment (SamMazerIR) [reply packages/spatial/src/input/systems/ClientInputSystem.tsx:478]: removed, wasn't sure if we wanted to rely on canvasEntity having the camera reference or if there was a separate way in some cases. will use canvasEntity now

File: packages/spatial/src/input/systems/ClientInputSystem.tsx (hunk 3 above)
Comment (SamMazerIR) [reply packages/spatial/src/input/systems/ClientInputSystem.tsx:140]: removed references to AvatarComponent

File: packages/spatial/src/input/systems/ClientInputSystem.tsx (hunk 2 above)
Comment (SamMazerIR) [reply packages/spatial/src/input/systems/ClientInputSystem.tsx:46]: removed references to AvatarComponent

File: packages/spatial/src/input/systems/ClientInputSystem.tsx (hunk 4 above)
Comment (SamMazerIR) [reply packages/spatial/src/input/systems/ClientInputSystem.tsx:294]: removed references to AvatarComponent

**Pull request merged**